drivers: comparator: mspm0: add a support for TI MSPM0 COMP module

Alex9360 opened the pull request:
Add driver support for comp module in mspm0 soc.

Signed-off-by: Alexpandi Muniyandi <alexpandi@linumiz.com>

Alex9360 committed:
dts: bindings: comparator: Add bindings for TI MSPM0 COMP Module

Add bindings for COMP modules in mspm0 g and l series

Signed-off-by: Alexpandi Muniyandi <alexpandi@linumiz.com>

Alex9360 committed:
drivers: comparator: mspm0: add support for TI MSPM0 COMP module

Add driver supporting the comp module in TI MSPM0 G and L series
with interrupt trigger and reference voltage support.

Add a window comparator mode feature for mspm0 g series

Signed-off-by: Alexpandi Muniyandi <alexpandi@linumiz.com>

Alex9360 committed:
dts: arm: ti: mspm0: Add COMP nodes for MSPM0 G and L series

Add a support for COMP nodes for mspm0 g and l series.

Signed-off-by: Alexpandi Muniyandi <alexpandi@linumiz.com>

ssekar15 commented on drivers/comparator/comparator_mspm0.c:
```diff
@@ -0,0 +1,341 @@
+	pending = DL_COMP_getEnabledInterruptStatus(config->regs,
+						    DL_COMP_INTERRUPT_OUTPUT_EDGE |
+						    DL_COMP_INTERRUPT_OUTPUT_EDGE_INV);
+	if (pending) {
```
if (pending != 0)

bjarki-andreasen reviewed (dismissed):
Looks good, do address the dtsi formatting errors though

Alex9360 commented:
@bjarki-andreasen dtsi errors solved. ping for review.

Alex9360 committed:
tests: drivers: comparator: gpio_loopback: Add overlay for test

Add board overlay for the TI LP-MSPM0G3507 to enable the
comparator gpio_loopback test using COMP0 with VDDA DAC
reference on the negative terminal.

Signed-off-by: Alexpandi Muniyandi <alexpandi@linumiz.com>

bjarki-andreasen commented on drivers/comparator/comparator_mspm0.c:
```diff
@@ -0,0 +1,343 @@
+	k_mutex_unlock(&data->dev_lock);
+	return ret;
+}
+
+static int comparator_mspm0_set_trigger_callback(const struct device *dev,
+						 comparator_callback_t callback,
+						 void *user_data)
+{
+	struct comparator_mspm0_data *data = dev->data;
+
+	k_mutex_lock(&data->dev_lock, K_FOREVER);
+	data->callback = callback;
+	data->user_data = user_data;
+	atomic_clear(&data->trigger_pending);
+	k_mutex_unlock(&data->dev_lock);
```
A mutex only blocks threads, not interrupts, so a callback can occur while the variables are being changed. This is bad if the callback is changed, then the interrupt occurs, since now the old user_data will be passed to the new callback.

To solve this best, disable the interrupt before changing the shared data, then enable it again.
```
        k_mutex_lock(&data->dev_lock, K_FOREVER);
        DL_COMP_disableInterrupt(...)
	data->callback = callback;
	data->user_data = user_data;
	atomic_clear(&data->trigger_pending);
        DL_COMP_enableInterrupt(...) /* only enable it again if it was already enabled */
	k_mutex_unlock(&data->dev_lock);
```

Alex9360 commented:
@bjarki-andreasen changes have been addressed. could you please review?

bjarki-andreasen commented:
Please address the compliance issues

Alex9360 commented:
@bjarki-andreasen ping for review. Thanks!

Alex9360 commented:
@ssekar15 @parthitce ping for review. Thanks!

bjarki-andreasen reviewed (changes requested):
The sample is not necessary given we have https://docs.zephyrproject.org/latest/hardware/peripherals/comparator.html#comparator-shell which can be used for playing around with the comparator.

Alex9360 commented:
@bjarki-andreasen removed the sample file.

bjarki-andreasen reviewed (dismissed):
Nice!

Alex9360 commented:
@ssekar15 ping for review. Thanks!

kartben commented on dts/bindings/comparator/ti,mspm0-comparator.yaml:
Please apply https://docs.zephyrproject.org/latest/build/dts/bindings-upstream.html#rules-for-default-values for all the default values

Alex9360 commented:
@kartben comment addressed!

Alex9360 commented:
@bjarki-andreasen ping for review. Thanks!
